File: find_shortest_string.rb
# Solution 1 
def find_shortest_string(arr)
  arr.reduce do |shortest, string|
    string.length < shortest.length ? string : shortest
  end
end

# Solution 2
# def find_shortest_string(arr)
#   shortest = arr[0]

#   arr.each do |string|
#     if string.length < shortest.length
#       shortest=string
#   end
# end
#   shortest
# end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  # Don't forget to add your own!

  # BENCHMARK HERE
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: find_shortest_string.js
function findShortestString(arr) {
  return arr.reduce((a,b) => a.length > b.length ?  b : a )
}

// a=shortest 
// b=string

 // THE LONGER WAY OF DOING THE EXACT SAME THING THAT'S ON LINE 2
//   let shortest = arr[0]

//   arr.forEach(string => {
//     if(string.length < shortest.length){
//       shortest = string
//     }
//   })
//   return shortest;
// }

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  // BENCHMARK HERE
  const longInput = [];

  for (let i = 0; i < 100; ++i) {
    longInput.push(Math.random());
  }

  const current = Date.now()

  for(let i=0; i<1000; i++){
    findShortestString(['flower', 'juniper', 'lily', 'dandelion']);
    findShortestString(longInput);
  }

  avgTime = (Date.now() - current) / 1000
  console.log(avgTime)
}

module.exports = findShortestString;

// Please add your pseudocode to this file
// And a written explanation of your solution
